File: backend/src/main/java/com/backend/petplace/domain/user/service/UserService.java
package com.backend.petplace.domain.user.service;

import com.backend.petplace.domain.email.entity.EmailAuthCode;
import com.backend.petplace.domain.email.repository.EmailAuthCodeRepository;
import com.backend.petplace.domain.user.dto.request.UserLoginRequest;
import com.backend.petplace.domain.user.dto.request.UserSignupRequest;
import com.backend.petplace.domain.user.dto.response.BoolResultResponse;
import com.backend.petplace.domain.user.dto.response.UserLoginResponse;
import com.backend.petplace.domain.user.dto.response.UserSignupResponse;
import com.backend.petplace.domain.user.entity.User;
import com.backend.petplace.domain.user.repository.UserRepository;
import com.backend.petplace.global.exception.BusinessException;
import com.backend.petplace.global.jwt.JwtTokenProvider;
import com.backend.petplace.global.response.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserService {

  private final UserRepository userRepository;
  private final PasswordEncoder passwordEncoder;
  private final JwtTokenProvider jwtTokenProvider;
  private final EmailAuthCodeRepository emailAuthCodeRepository;

  @Transactional
  public UserSignupResponse signup(UserSignupRequest request) {
     validateDuplicateNickName(request.getNickName());

     validateDuplicateEmail(request.getEmail());

    // 이메일 인증 체크
    checkAuthCode(request);

    User user = User.create(request, passwordEncoder.encode(request.getPassword()));
    userRepository.save(user);

    return new UserSignupResponse(user.getId());
  }

  @Transactional
  protected void checkAuthCode(UserSignupRequest request) {
    EmailAuthCode emailAuthCode = emailAuthCodeRepository.findByEmailAndAuthCode
            (request.getEmail(), request.getAuthCode())
        .orElseThrow(() -> new BusinessException(ErrorCode.AUTH_CODE_NOT_FOUND));

    if (emailAuthCode.isVerified()) {
      emailAuthCodeRepository.delete(emailAuthCode);
      return;
    }
    throw new BusinessException(ErrorCode.AUTH_CODE_NOT_VERIFIED);
  }

  @Transactional(readOnly = true)
  public BoolResultResponse validateDuplicateNickName(String nickName) {
    if (userRepository.existsByNickName(nickName)) {
      throw new BusinessException(ErrorCode.DUPLICATE_NICKNAME);
    }
    return new BoolResultResponse(true);
  }

  @Transactional(readOnly = true)
  public BoolResultResponse validateDuplicateEmail(String email) {
    if (userRepository.existsByEmail(email)) {
      throw new BusinessException(ErrorCode.DUPLICATE_EMAIL);
    }
    return new BoolResultResponse(true);
  }

  @Transactional(readOnly = true)
  public UserLoginResponse login(UserLoginRequest request) {
    User user = userRepository.findByNickName(request.getNickName())
        .orElseThrow(() -> new BusinessException(ErrorCode.BAD_CREDENTIAL));

    if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
      throw new BusinessException(ErrorCode.BAD_CREDENTIAL);
    }

    return new UserLoginResponse(jwtTokenProvider.generateAccessToken(user.getId()));
  }
}

File: backend/src/main/java/com/backend/petplace/domain/user/service/UserService.kt
package com.backend.petplace.domain.user.service

import com.backend.petplace.domain.email.repository.EmailAuthCodeRepository
import com.backend.petplace.domain.user.dto.request.UserLoginRequest
import com.backend.petplace.domain.user.dto.request.UserSignupRequest
import com.backend.petplace.domain.user.dto.response.BoolResultResponse
import com.backend.petplace.domain.user.dto.response.UserLoginResponse
import com.backend.petplace.domain.user.dto.response.UserSignupResponse
import com.backend.petplace.domain.user.entity.User
import com.backend.petplace.domain.user.repository.UserRepository
import com.backend.petplace.global.exception.BusinessException
import com.backend.petplace.global.jwt.JwtTokenProvider
import com.backend.petplace.global.response.ErrorCode
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtTokenProvider: JwtTokenProvider,
    private val emailAuthCodeRepository: EmailAuthCodeRepository
) {
    @Transactional
    fun signup(request: UserSignupRequest): UserSignupResponse {
        validateDuplicateNickName(requireNotNull(request.nickName))
        validateDuplicateEmail(requireNotNull(request.email))

        checkAuthCode(request)

        val user = User.create(request, passwordEncoder.encode(request.password))
        userRepository.save(user)

        return UserSignupResponse(requireNotNull(user.id))
    }

    @Transactional
    protected fun checkAuthCode(request: UserSignupRequest) {
        val emailAuthCode = emailAuthCodeRepository.findByEmailAndAuthCode(
            requireNotNull(request.email),
            requireNotNull(request.authCode)
        ) ?: throw BusinessException(ErrorCode.AUTH_CODE_NOT_FOUND)

        if (emailAuthCode.verified) {
            emailAuthCodeRepository.delete(emailAuthCode)
            return
        }
        throw BusinessException(ErrorCode.AUTH_CODE_NOT_VERIFIED)
    }

    @Transactional(readOnly = true)
    fun validateDuplicateNickName(nickName: String): BoolResultResponse {
        if (userRepository.existsByNickName(nickName)) {
            throw BusinessException(ErrorCode.DUPLICATE_NICKNAME)
        }
        return BoolResultResponse(true)
    }

    @Transactional(readOnly = true)
    fun validateDuplicateEmail(email: String): BoolResultResponse {
        if (userRepository.existsByEmail(email)) {
            throw BusinessException(ErrorCode.DUPLICATE_EMAIL)
        }
        return BoolResultResponse(true)
    }

    @Transactional(readOnly = true)
    fun login(request: UserLoginRequest): UserLoginResponse {
        val user = userRepository.findByNickName(requireNotNull(request.nickName))
            ?: throw BusinessException(ErrorCode.BAD_CREDENTIAL)

        if (!passwordEncoder.matches(request.password, user.password)) {
            throw BusinessException(ErrorCode.BAD_CREDENTIAL)
        }

        return UserLoginResponse(jwtTokenProvider.generateAccessToken(user.id!!))
    }
}